### ISSUE TITLE
Graph partitioning interface

### ISSUE THREAD

**Issue by mtfishman** [mtfishman/ITensorNetworks.jl#28]
Currently we have functions:
```julia
using NamedGraphs
using ITensors
using ITensorNetworks
using Metis

tn = ITensorNetwork(named_grid((4, 4)); link_space=4)
partition(tn, 2)
subgraphs(tn, nv(tn) ÷ 2)
```
for partitioning graphs, which output:
```julia
julia> partition(tn, 2)
16-element Dictionaries.Dictionary{Tuple{Int64, Int64}, Int64}
 (1, 1) │ 1
 (2, 1) │ 1
 (3, 1) │ 1
 (4, 1) │ 1
 (1, 2) │ 1
 (2, 2) │ 1
 (3, 2) │ 1
 (4, 2) │ 1
 (1, 3) │ 2
 (2, 3) │ 2
 (3, 3) │ 2
 (4, 3) │ 2
 (1, 4) │ 2
 (2, 4) │ 2
 (3, 4) │ 2
 (4, 4) │ 2

julia> subgraphs(tn, nv(tn) ÷ 2)
DataGraph{Int64, Vector{Tuple{Int64, Int64}}, Any, NamedGraph{Int64}, NamedEdge{Int64}} with 2 vertices:
2-element Vector{Int64}:
 1
 2

and 1 edge(s):
1 => 2

with vertex data:
2-element Dictionaries.Dictionary{Int64, Vector{Tuple{Int64, Int64}}}
 1 │ [(1, 1), (2, 1), (3, 1), (4, 1), (1, 2), (2, 2), (3, 2), (4, 2)]
 2 │ [(1, 3), (2, 3), (3, 3), (4, 3), (1, 4), (2, 4), (3, 4), (4, 4)]

and edge data:
0-element Dictionaries.Dictionary{NamedEdge{Int64}, Any}
```
We should make them consistent so that they both can accept the partition size or the number of partitions. Additionally, their names should give a better indication for what they output.

For example, I could imagine a function:
```julia
julia> partition_vertices(tn, 2)
2-element Vector{Vector{Tuple{Int64, Int64}}}:
 [(1, 1), (2, 1), (3, 1), (4, 1), (1, 2), (2, 2), (3, 2), (4, 2)]
 [(1, 3), (2, 3), (3, 3), (4, 3), (1, 4), (2, 4), (3, 4), (4, 4)]

julia> partition_vertices(tn; partition_size=nv(tn) ÷ 2)
2-element Vector{Vector{Tuple{Int64, Int64}}}:
 [(1, 1), (2, 1), (3, 1), (4, 1), (1, 2), (2, 2), (3, 2), (4, 2)]
 [(1, 3), (2, 3), (3, 3), (4, 3), (1, 4), (2, 4), (3, 4), (4, 4)]
```
Or perhaps `partition_vertices` should output a `DataGraph` like `subgraphs` does right now, which also keeps track of which partitions have connected edges.

@JoeyT1994 curious about your opinion on this.

**Comment by emstoudenmire** [comment]
Agreed the new function interface is better. Would it make sense in the future to allow partitioning by properties too? Like a keyword argument `by=...` that takes a function that determines which partition to put a vertex into? (Only if there's a use case for it of course.)

**Comment by mtfishman** [comment]
That kind of thing would be easy enough to do using the `SplitApplyCombine.group*` functions on `vertices(tn)`: https://github.com/JuliaData/SplitApplyCombine.jl#grouping. These functions are more focused on graph partitioning based on graph properties, and wrap functionality from packages like [Metis.jl](https://github.com/JuliaSparse/Metis.jl) and [KaHyPar.jl](https://github.com/kahypar/KaHyPar.jl).

**Comment by mtfishman** [comment]
I could imagine a functionality like:
```julia
subgraph_vertices = [
  [(1, 1), (2, 1), (3, 1), (4, 1), (1, 2), (2, 2), (3, 2), (4, 2)],
  [(1, 3), (2, 3), (3, 3), (4, 3), (1, 4), (2, 4), (3, 4), (4, 4)],
]
dg = subgraph_connectivity(tn, subgraph_vertices)
```
which outputs a graph with edges indicating which subgraphs are connected to each other (like the output of the currently available `subgraphs` function I show above), in this case an edge `1 => 2` indicating subgraph `1` has at least one edge connected to subgraph `2`. The `subgraph_vertices` input could then be determined using functions like `partition_vertices` or `SplitApplyCombine.group`.

**Comment by JoeyT1994** [comment]
Yes I think that a `partition_vertices' function which allows the user to choose partition_number vs partition_size would be helpful. 

@emstoudenmire, there are a few cases from BP I can think of where partitioning by properties would be helpful..

For instance if we want to do belief propagation over a wave function `tn` but don't want to explicitly flatten the network due to memory overhead we can form Z = <tn|tn> without actually doing the contraction via

<img width="541" alt="image" src="https://user-images.githubusercontent.com/51231103/208934790-6d5fa0b1-e30a-4aef-8bbf-3571888e1e80.png">

It is then most logical to form ensure those ITensors in Z which share a site index are placed into the same subgraphs and do belief propagation that way. 

Additionally if we do want to efficiently do a Belief Propagation version of DMRG we would need to form (but not contract) <tn|H|tn> and again want to keep all ITensors which share site inds in the same subgraphs. 

@mtfishman In the case of Simple Belief Propagation this seems quite easily done with `SplitApplyCombine.group` (as those tensors which share a site index have all elements of their vertex tuple identical - other than the first). What, however,  if I wanted to have larger subgraphs for GBP (e.g. 2*n ITensors of Z in each subgraph, each ITensor must share the subgraph with its conjugate ITensor, and the network of ITensors formed within each subgraph must be connected)?

Perhaps we would need to write up a function for that?

With this functionality the code we have written in `beliefpropagation.jl` can be used to form SBP or GBP message tensors of such objects without needing any keyword arguments itself, as all of the subgraphing is taken care of elsewhere which is neat.

**Comment by JoeyT1994** [comment]
Hi Matt,

Thanks to the solution to that problem. That is really neat and I agree that writing some helper functions that can turn it into a few lines when it is needed to be used mid-program would be perfect.

Do you think such functions should go into `subgraphs.jl`? For instance one could be `subgraph_connectivity` as you wrote above, with the extra that the edge data is number of edges from the original graph actually involved. 

Tomorrow, I could add that to `subgraphs.jl` and make a PR if you wish?

**Comment by mtfishman** [comment]
Closed by #30.
